pool_stream.go

```diff
@@ -0,0 +1,172 @@
+package dht
+
+import (
+	"context"
+	"log"
+	"sync"
+
+	pbio "github.com/gogo/protobuf/io"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"golang.org/x/xerrors"
+)
+
+func (dht *IpfsDHT) getPoolStream(ctx context.Context, p peer.ID) (*poolStream, error) {
+	dht.streamPoolMu.Lock()
+	for ps := range dht.streamPool[p] {
+		dht.deletePoolStreamLocked(ps, p)
+		if ps.bad() {
+			log.Printf("got bad pool stream for %v", p)
+			continue
+		}
+		dht.streamPoolMu.Unlock()
+		log.Printf("reusing pool stream for %v", p)
+		return ps, nil
+	}
+	dht.streamPoolMu.Unlock()
+	log.Printf("creating new pool stream for %v", p)
+	return dht.newPoolStream(ctx, p)
+}
+
+func (dht *IpfsDHT) putPoolStream(ps *poolStream, p peer.ID) {
+	dht.streamPoolMu.Lock()
+	defer dht.streamPoolMu.Unlock()
+	if ps.bad() {
+		log.Printf("putting pool stream for %v but it went bad", p)
+		return
+	}
+	log.Printf("putting pool stream for %v", p)
+	if dht.streamPool[p] == nil {
+		dht.streamPool[p] = make(map[*poolStream]struct{})
+	}
+	dht.streamPool[p][ps] = struct{}{}
+}
+
+func (dht *IpfsDHT) deletePoolStream(ps *poolStream, p peer.ID) {
+	dht.streamPoolMu.Lock()
+	dht.deletePoolStreamLocked(ps, p)
+	dht.streamPoolMu.Unlock()
+}
+
+func (dht *IpfsDHT) deletePoolStreamLocked(ps *poolStream, p peer.ID) {
+	log.Printf("deleting pool stream for %v", p)
+	delete(dht.streamPool[p], ps)
+	if len(dht.streamPool) == 0 {
+		delete(dht.streamPool, p)
+	}
+}
+
+func (dht *IpfsDHT) newPoolStream(ctx context.Context, p peer.ID) (*poolStream, error) {
+	s, err := dht.newStream(ctx, p)
+	if err != nil {
+		return nil, xerrors.Errorf("opening stream: %w", err)
+	}
+	ps := &poolStream{
+		stream: s,
+		w:      newBufferedDelimitedWriter(s),
+		r:      pbio.NewDelimitedReader(s, inet.MessageSizeMax),
+		m:      make(chan chan *pb.Message, 1),
+	}
+	ps.onReaderErr = func() {
+		ps.reset()
+		dht.deletePoolStream(ps, p)
+	}
+	go ps.reader()
+	return ps, nil
+}
+
+type poolStream struct {
+	stream interface {
+		Reset() error
+	}
+	w           bufferedWriteCloser
```

Comment:
I realize these are the old names, but no reason not to improve now. Unless these are idiomatic in go?

Reply:
It is quite idiomatic in Go, here it would be clear that `w` is the thing that does the writing for `poolStream`. I'm open to suggestions for better names though, if you can think of a more appropriate one.